### PR TITLE
Run namespaces in their own thread pool

### LINKED ISSUE #71: Locks up (usually) when there are more namespaces than threads
This happens when "multithread namespaces" is on.

In VisualVM, I see that all executer threads are parked, waiting on `.get` of `FutureTask`.

Given the stack traces, I'm confident that the problem is that every executer thread is trying to run a namespace in parallel, submitting back to the executer, but there are no threads left in the executer to run tests.

Note that this means that a fair number of threads are idle while running the tests.

Work-around: for now, disable multithread namespaces.
A possible solution: Use a separate thread pool with no limit `(Executors/newCachedThreadPool)` for running namespaces.  Or somehow coordinate differently?

### PR DESCRIPTION
**Pull request by eraserhd**
Closes #71

Except for setting up and tearing down "once fixtures", namespace threads are
parked, meaning these cores aren't available for running tests.  Additionally,
when there are less threads than namespaces, the runner would deadlock when
all test executor threads were being used by namespaces waiting for tests.

### CODE REVIEW

File: eftest/src/eftest/runner.clj [hunk 1]
```diff
@@ -48,7 +48,7 @@
         (with-bindings* bindings f)))))
 
 (defn- default-thread-count []
-  (+ 2 (.availableProcessors (Runtime/getRuntime))))
+  (.availableProcessors (Runtime/getRuntime)))
```
Comment (weavejester) [inline eftest/src/eftest/runner.clj]: Why is this change necessary?

File: eftest/src/eftest/runner.clj [hunk 2]
```diff
@@ -120,18 +120,22 @@
...
-                 (partial pmap* executor)
-                 map)
-        f      #(->> (group-by (comp :ns meta) vars)
-                     (mapf (fn [[ns vars]] (test-ns ns vars report opts)))
-                     (apply merge-with +))]
-    (if capture-output?
-      (capture/with-capture (f))
-      (f))))
+  (let [report             (synchronize test/report)
+        namespace-executor (when (multithread-namespaces? opts)
```
Comment (weavejester) [inline eftest/src/eftest/runner.clj]: Change the name to `executor`. Too much whitespace makes it harder to read the other bindings in the `let`.

Also, rather than using `when`, a `delay` might be a better option here, e.g.

```
(delay (Executors/newCachedThreadPool))
```

File: eftest/src/eftest/runner.clj (hunk 1 above)
Comment (eraserhd) [reply eftest/src/eftest/runner.clj]: I *presumed* this was an attempt to account for namespace threads.  I can revert if this is not true.

**Comment by eraserhd** [comment]
@weavejester I extracted a protocol to remove several conditionals instead of using `delay`.  If that doesn't look better to you, I'll redo it.

Not entirely comfortable with the protocol name.

And, let me know if I should squash the commits.

Thanks!
-Jason

File: eftest/src/eftest/runner.clj (hunk 1 above)
Comment (weavejester) [reply eftest/src/eftest/runner.clj]: It uses the same calculation as Clojure does. My assumption is that it does this because by adding 2, you guarantee that there are at least that number of threads. It might be a protection against an incorrect value returned by `.availableProcessors`.

File: eftest/src/eftest/runner.clj [hunk 3]
```diff
@@ -8,6 +8,9 @@
             [eftest.output-capture :as capture])
   (:import [java.util.concurrent Executors ExecutorService]))
 
+(defprotocol Executor
```
Comment (weavejester) [inline eftest/src/eftest/runner.clj]: I don't understand the reason for adding this protocol. It seems to use more code and indirection to achieve the same result as before.

File: eftest/src/eftest/runner.clj (hunk 2 above)
Comment (eraserhd) [reply eftest/src/eftest/runner.clj]: ok, this is all done

**Comment by eraserhd** [comment]
That's all feedback addressed @weavejester

**Comment by weavejester** [comment]
Thanks for the update. Can you reformat the commit message to:

```
Run namespaces in their own thread pool

Except for setting up and tearing down "once fixtures", namespace
threads are parked, meaning these cores aren't available for running
tests. Additionally, when there are less threads than namespaces, the
runner would deadlock when all test executor threads were being used by
namespaces waiting for tests.

Closes #71.
```

So a wordwrap at the more standard 72 characters rather than 80 for the commit body.

Then I should be able to merge.

**Comment by eraserhd** [comment]
@weavejester done

**Comment by weavejester** [comment]
Manually merged.
